content/blog/2019-12-18-novm/index.md

```diff
@@ -0,0 +1,317 @@
+
+## Overhead of Stack Checking
+
+We evaluated the performance impact of using split stacks on two microbenchmarks
+designed to be bottlenecked on function calls, which respectively do and do not
+trigger run-time memory allocations.
+The first microbenchmark was a naive program to compute the 50th
+number in the fibonnacci sequence without memoization; this did not require
+a large amount of stack so we use it to measure the overhead of _just checking_
+whether or not there is enough space.
+The other microbenchmark naively recursively computes the sum of the first _n_ integers.
+We executed this benchmark with `n=1000000000` so that it would trigger run-time allocations
+by recursing very deeply.
+
+<img src="stackcheckmicro.png" style="width:100%"/>
+
+
+This diagram plots the execution time of these two microbenchmarks with the `-fstack-split`
+option enabled, normalized to regular execution (statically allocated stacks).
+As you can see, our fibonnacci benchmark has only about a 15% increase in runtime
+caused by checking remaining stack space. While not an insignificant cost, most
+programs will not execute nearly as high a density of function calls and should not
+see such high overheads.
+The recursive sum benchmark ended up executing 6410250 different run-time allocations
+of 1544 bytes. While it had a very large performance impact, we could certainly tune
+the stack allocation algorithm to request larger chunks of memory to reduce the frequency
+of `malloc` system calls.
+
+### PARSEC Benchmarks
+
+While these microbenchmarks give us a good upper bound on worst-case overhead,
+we wanted to evaluate some more realistic tests. We chose the [PARSEC](https://parsec.cs.princeton.edu/)
+benchmarks, mostly because we used them for a prior project in this class and could test them easily.
+The execution times for these benchmarks are the built-in "Regions of Interest" and exclude initialization
+and warm up times for each program.
+
+<img src="stackcheckparsec.png" style="width:100%"/>
+
+With these four benchmarks, there was almost no impact of applying the split-stack option.
+We should note that the StreamCluster benchmark actually _sped up_ when using split-stack;
+likely this is some sort of memory alignment effect à la [Mytcowicz et al.](https://dl.acm.org/citation.cfm?id=1508275).
+In any case, we should probably consider this impact to be negligible.
+
+Of these benchmarks, only Ferret actually required dynamically allocating stack space.
+Each of these allocations was 618 KB, which is a potential concern. It is unclear, in a real
+system using only physical addressing, whether or not allocations of this size would
+be frequently servicable or not. I hypothesize that real systems with many gigabytes or terabytes
+of memory with even severe fragmentation will be able to regularly respond to allocations
+in the kilobyte range; however, proving this is future work.
```

Comment:
It does seem like such a futuristic system would probably want to hand out memory to applications in largish chunks. 4 KB in particular seems like too small an increment to allocate to a modern application at a time. I can see this working by handing out biggish chunks, even when the application doesn't need them, and then letting userspace manage the fragmentation within those chunks.